src/serializer/ResidualHeapVisitor.js

```diff
@@ -383,8 +394,33 @@ export class ResidualHeapVisitor {
         if (functionInfo.modified.has(innerName)) residualFunctionBinding.modified = true;
       }
     });
-
+    let useClassProperties = false;
+    let isConstructor = false;
+    let isComputed = !!val.$HasComputedName;
+
+    if (val.$HomeObject instanceof ObjectValue) {
+      // determine if our $HomeObject is actually a ES2015 class by looking at constructor
+      let constructorValue = Get(this.realm, val.$HomeObject, "constructor");
+      if (constructorValue instanceof AbstractValue) {
+        throw new FatalError("TODO: do not know how to handle an abstract constructor");
+      } else if (constructorValue instanceof ECMAScriptSourceFunctionValue) {
+        if (constructorValue.$FunctionKind === "classConstructor") {
+          useClassProperties = true;
+          if (val.$FunctionKind === "classConstructor") {
+            isConstructor = true;
+          }
+        }
+      }
+    }
     this.functionInstances.set(val, {
+      classProperties: useClassProperties
```

Comment:
The name suggests to me that it would contain a list of properties. Maybe call it something like `methodData` or something?